File: frontend/src/services/candidate.ts
import axios from 'axios';
import { API_URL } from './auth';

export interface Course {
  id: string;
  title: string;
  description: string;
  pdf_url: string;
}

export interface Exam {
  id: string;
  subject_id: string;
  title: string;
  csv_url: string;
}

export interface ExamAttempt {
  id: string;
  candidate_id: string;
  exam_id: string;
  score_percentage: number;
  passed: boolean;
  attempted_on: string;
}

export interface ExamSubmission {
  exam_id: string;
  answers: Record<string, string>;
}

export interface ExamResult {
  score_percentage: number;
  passed: boolean;
}

export interface CourseCertificate {
  id: string;
  candidate_id: string;
  course_id: string;
  certificate_url: string;
  issued_on: string;
}

export interface Progress {
  total_courses: number;
  total_subjects: number;
  completed_subjects: number;
  earned_certificates: number;
  completion_percentage: number;
}

export interface ExamQuestion {
  question: string;
  option_a: string;
  option_b: string;
  option_c: string;
  option_d: string;
}

export interface PaginatedExamQuestions {
  questions: ExamQuestion[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export const candidateService = {
  async getAvailableCourses(): Promise<Course[]> {
    const response = await axios.get(`${API_URL.replace('/auth', '')}/candidate/courses`);
    return response.data;
  },

  async getAvailableExams(): Promise<Exam[]> {
    const response = await axios.get(`${API_URL.replace('/auth', '')}/candidate/exams`);
    return response.data;
  },

  async submitExam(submission: ExamSubmission): Promise<ExamResult> {
    const response = await axios.post(`${API_URL.replace('/auth', '')}/candidate/exams/submit`, submission);
    return response.data;
  },

  async getExamAttempts(): Promise<ExamAttempt[]> {
    const response = await axios.get(`${API_URL.replace('/auth', '')}/candidate/attempts`);
    return response.data;
  },

  async getCertificates(): Promise<CourseCertificate[]> {
    const response = await axios.get(`${API_URL.replace('/auth', '')}/candidate/certificates`);
    return response.data;
  },

  async getProgress(): Promise<Progress> {
    const response = await axios.get(`${API_URL.replace('/auth', '')}/candidate/progress`);
    return response.data;
  },

  async getExamQuestions(
    examId: string,
    page: number = 1,
    page_size: number = 10
  ): Promise<PaginatedExamQuestions> {
    const response = await axios.get<PaginatedExamQuestions>(
      `${API_URL.replace('/auth', '')}/candidate/exams/${examId}/questions`,
      {
        params: { page, page_size },
        headers: { Authorization: `Bearer ${localStorage.getItem('token')}` }
      }
    );
    return response.data;
  }
}; 

File: backend/backend/routers/candidate.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List
from uuid import UUID

from backend.core.dependencies import get_current_user, get_db
from backend.models.user import User
from backend.schemas.exam import (
    ExamInDB, ExamAttemptInDB, ExamSubmission, ExamResult,
    ExamQuestionResponse, PaginatedExamQuestions
)
from backend.schemas.course import CourseCertificateInDB
from backend.services import candidate as candidate_service

router = APIRouter()

@router.get("/courses", response_model=List[dict])
async def list_available_courses(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all available courses"""
    return candidate_service.get_available_courses(db)

@router.get("/exams", response_model=List[ExamInDB])
async def list_available_exams(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all available exams for the candidate"""
    return candidate_service.get_available_exams(db, current_user.id)

@router.post("/exams/submit", response_model=ExamResult)
async def submit_exam(
    submission: ExamSubmission,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Submit an exam attempt"""
    return candidate_service.submit_exam(db, submission, current_user.id)

@router.get("/attempts", response_model=List[ExamAttemptInDB])
async def list_exam_attempts(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all exam attempts for the candidate"""
    return candidate_service.get_exam_attempts(db, current_user.id)

@router.get("/certificates", response_model=List[CourseCertificateInDB])
async def list_certificates(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all certificates earned by the candidate"""
    return candidate_service.get_certificates(db, current_user.id)

@router.get("/progress", response_model=dict)
async def get_progress(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get candidate's progress across courses"""
    return candidate_service.get_candidate_progress(db, current_user.id)

@router.get("/exams/{exam_id}/questions", response_model=PaginatedExamQuestions)
async def get_exam_questions(
    exam_id: UUID,
    page: int = Query(1, ge=1, description="Page number (1-based)"),
    page_size: int = Query(10, ge=1, le=50, description="Number of questions per page"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get paginated questions for an exam without correct answers"""
    return candidate_service.get_exam_questions(db, exam_id, page, page_size)
